include/lib/map.h

```diff
@@ -4,24 +4,39 @@
 
 #include "stdlib/stdtypes.h"
 
-/** Data structure for a generic map */
+/** Structures for a generic map. */
+struct entry {
+    /** The key associated with the entry. */
+    void *key;
+    /** The value associated with the entry. */
+    void *value;
+};
+
 struct map {
-    //TODO
+    /** Table which holds all entries. */
+    struct entry *table;
+    /** Amount of space used. */
+    size_t used;
+    /** Maximum space to allocate for entries */
+    size_t size;
 };
 
-/** Initializes a new hashtable */
-void map_init(struct map *m);
+/** Initializes a new hashmap. Returns `-1` if not allocated. */
+int map_init(struct map *m, size_t size);
 
-/** Frees the hashtable from memory */
+/** Frees the hashtable from memory. */
 void map_free(struct map *m);
 
-/** Adds the key/value pair to the hashtable */
-void map_put(struct map *m, void *key, void *value);
+/** Adds the key/value pair to the hashtable. */
+int map_put(struct map *m, void *key, void *value, size_t size);
```

Comment:
There is no reason for this function to return an int. You should keep this a void function